contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,213 @@
+
+## Function Detail
+
+### API Definition
+
+The following requirements are identified during the discussion of this feature:
+
+1. Existing workload can use this feature without (or with a few) configuration changes
+2. Pods of a group/gang may have different `PodSpec` (and/or belong to different collections)
+3. Existing controllers which are responsible for managing life cycle of collections work well with this feature
+
+To meet the requirements above, the following **Kind** is introduced in core part under `scheduling.k8s.io/v1alpha1` **Group**/**Version**.
+
+```go
+// PodGroup defines the scheduling requirement of a pod group
+type PodGroup struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+	// Spec defines the behavior of a pod group.
+	// +optional
+    Spec PodGroupSpec
+
+	// Status represents the current information about a pod group.
+	// This data may not be up to date.
+	// +optional
+    Status PodGroupStatus
+}
+
+// LifecyclePolicy represents the lifecycle policy of PodGroup.
+type LifeCyclePolicy struct {
+    // The action that will be taken to the PodGroup according to Event.
+    // One of "Restart", "None".
+    // Default to None.
+    // +optional
+    Action Action
+
+    // The Event recorded by scheduler; the controller takes actions
+    // according to this Event.
+    // One of "PodFailed", "Unschedulable".
+    // +optional
+    Event Event
+
+    // Timeout is the grace period for controller to take actions.
+    // Default to nil (take action immediately).
+    // +optional
+    Timeout *metav1.Duration
+}
+
+// PodGroupSpec represents the template of a pod group.
+type PodGroupSpec struct {
+    // NumMember defines the number of members/tasks to run the pod group;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    NumMember int
+
+    // Policy defines the policy of PodGroup lifecycle.
+    // Default to 'Action: None, PodPhase: Failed'
+    // +optional
+    Policy []LifeCyclePolicy
+}
+
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+    // The number of actively running pods.
+    // +optional
+    Running int32
+
+    // The number of pods which reached phase Succeeded.
+    // +optional
+    Succeeded int32
+
+    // The number of pods which reached phase Failed.
+    // +optional
+    Failed int32
+}
+```
+
+The `PodGroup`, which is a namespaced object, specifies the behavior and status of a pod group, e.g. number of pods in a group, lifecycle policy. To define which pods are member of `PodGroup`, the following field is introduced to `PodSpec`.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+    ...
+
+    // The name of PodGroup that the pod belongs to; the pod
+    // must belong to the PodGroup in the same namespace.
+    // +optional
+    GroupName string
```

Comment:
[feedback from api reviewers](https://github.com/kubernetes/community/pull/2337#discussion_r203909088) was that this could/should be modeled as a label rather than a spec field. I don't understand the point about a race condition raised in https://github.com/kubernetes/community/pull/2337#discussion_r203954962... a `metadata.labels['scheduling.k8s.io/group-name']` label on a pod carries just as much information as a `spec.groupName` field, and has the additional benefit of allowing label-selection queries to filter pods to a specific group name

Reply:
@liggitt 
I like the flexibility of this new pattern, but aren't system components supposed to not look for specific label keys for their core functionalities?

Reply:
I thought the scheduler acting on this was being built out of tree

Reply:
Done :)